### ISSUE TITLE
Make composition of asynchronous actions easier

### ISSUE THREAD

**Issue by mikesname** [playframework/playframework#1827]
Since 2.2, actions can no longer yield an `AsyncResult`, but instead must use a `Future[PlainResult]`. This means that composing actions which run in a deferred context (i.e. by fetching some data from a asynchronous data source) is more difficult, since users can no longer simply return a `Result` and rely on the [recursive unflattening](https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/scala/play/api/mvc/Results.scala#L471) of nested `AsyncResult`s that used to occur. 

Consider the following example. If I am using an asynchronous data source (a ReST service accessed via `WS`, for example), and for _every request_ I was to fetch the currently logged-in user's profile and permissions info I might have done this:

``` scala
  def UserProfileAction(f: Option[UserProfile] => Request[AnyContent] => Result): Action[AnyContent] = {
    Action { request =>
      request.session.get("user").map { userId => // simplification here!!!
        Async {
          // some ReST calls with userId
          val profileRequest: Future[UserProfile] = ???
          val permissionsRequest: Future[GlobalPermissionSet[UserProfile]] = ???
          for {
            profile <- profileRequest
            permissions <- permissionsRequest
          } yield {
            f(Some(profile.copy(globalPermissions = Some(permissions))))(request)
          }
        }
      } getOrElse {
        f(None)(request)
      }
    }
  }
```

If I then wanted to compose this action with one that fetched item data and item-level permissions for the given user I'd be able to do:

``` scala
  def ItemAction(id: String)[T](f: T => Option[UserProfile] => Request[AnyContent] => Result)(
      implicit reader: ItemReader[T]) = {
    UserProfileAction { implicit userOpt => implicit request =>
      Async {
        // some ReST calls depending on the user profile info
        val itemRequest: Future[T] = ???
        val itemPermsRequest: Future[ItemPermissionSet[UserProfile]] = ???
        for {
          item <- itemRequest
          itemPerms <- itemPermsRequest
        } yield {
          f(item)(userOpt.map(_.copy(itemPermissions = Some(itemPerms))))(request)
        }
      }
    }
  }
```

It might not be pretty (or massively efficient - since the ReST calls aren't all in parallel), but it made for nice controller actions that just looked something like this:

``` scala
object ThingController extends Controller with MyActions[Thing] {
  def showThing(id: String) = ItemAction(id) { 
      item => implicit userOpt => implicit request =>
    Ok(views.html.show(item))
  }
}
```

The eventual `Result` from this action might be nested in several futures (though in practice much of the data is retrieved straight from the cache.) My actions are also parametrized and handle more complex CRUD operations for many different item types via implicit `Reader[T]` and `Writer[T]` objects. The main point is that never having to worry about how deferred the eventual `Result` was was (previously) a _massive bonus_.

This scheme is proving difficult to reproduce with the 2.2 changes to the `Result` structure:
- There's a simple `Action.async` helper to do asynchronous operations but since the result must be a `Future[PlainResult]` unflattening of nested futures must be done manually
- Custom actions which do asynchronous operations can be constructed with `ActionBuilder`, but that appears to be [awkward to compose](https://github.com/mariussoutier/PlayBasics/blob/master/play-2.2-migration/app/controllers/ActionComposition.scala#L139)
- `ActionBuilder` appears to be difficult to parametrize (or use with a parametrized `WrappedRequest`) unless nested within a parametrized outer class/object

I'd be very happy if this situation could be "fixed" via enhancements to the documentation that'd make it clear how to achieve what I currently do with Play 2.1.x. My current impression is that the framework has taken a backward step with 2.2 by assuming that only one level of `Future` is ever needed, when in practice if your main data source is an asynchronous one the situation is somewhat more complex.
